packages/ra-ui-materialui/src/button/ExportButton.tsx

```diff
@@ -0,0 +1,170 @@
+import React, { useCallback, FunctionComponent } from 'react';
+import PropTypes from 'prop-types';
+import DownloadIcon from '@material-ui/icons/GetApp';
+import {
+    downloadCSV,
+    useDataProvider,
+    useNotify,
+    GET_MANY,
+    GET_LIST,
+    Sort,
+    DataProvider,
+} from 'ra-core';
+import jsonExport from 'jsonexport/dist';
+
+import Button from './Button';
+
+const sanitizeRestProps = ({ basePath, ...rest }: any) => rest;
+
+/**
+ * Extracts, aggregates and deduplicates the ids of related records
+ *
+ * @example
+ *     const books = [
+ *         { id: 1, author_id: 123, title: 'Pride and Prejudice' },
+ *         { id: 2, author_id: 123, title: 'Sense and Sensibility' },
+ *         { id: 3, author_id: 456, title: 'War and Peace' },
+ *     ];
+ *     getRelatedIds(books, 'author_id'); => [123, 456]
+ *
+ * @example
+ *     const books = [
+ *         { id: 1, tag_ids: [1, 2], title: 'Pride and Prejudice' },
+ *         { id: 2, tag_ids: [2, 3], title: 'Sense and Sensibility' },
+ *         { id: 3, tag_ids: [4], title: 'War and Peace' },
+ *     ];
+ *     getRelatedIds(records, 'tag_ids'); => [1, 2, 3, 4]
+ *
+ * @param {Object[]} records An array of records
+ * @param {string} field the identifier of the record field to use
+ */
+export const getRelatedIds = (records, field) =>
+    Array.from(
+        new Set(
+            records
+                .filter(record => record[field] != null)
+                .map(record => record[field])
+                .reduce((ids, value) => ids.concat(value), [])
+        )
+    );
+
+/**
+ * Helper function for calling the data provider with GET_MANY
+ * via redux and saga, and getting a Promise in return
+ *
+ * @example
+ *     fetchRelatedRecords(records, 'post_id', 'posts').then(posts =>
+ *          posts.map(record => ({
+ *              ...record,
+ *              post_title: posts[record.post_id].title,
+ *          }));
+ */
+const fetchRelatedRecords = dataProvider => (data, field, resource) =>
+    dataProvider(GET_MANY, resource, { ids: getRelatedIds(data, field) }).then(
+        ({ data }) =>
+            data.reduce((acc, post) => {
+                acc[post.id] = post;
+                return acc;
+            }, {})
+    );
+
+const DefaultIcon = <DownloadIcon />;
+const defaultFilter = {};
+
+interface Props {
+    exporter?: (
+        data: any,
+        fetchRelatedRecords: (
+            data: any,
+            field: string,
+            resource: string
+        ) => Promise<any>,
+        dataProvider: DataProvider
+    ) => Promise<void>;
+    sort: Sort;
+    filter?: any;
+    maxResults: number;
+    resource: string;
+    onClick?: (e: Event) => void;
+    label: string;
+    icon?: JSX.Element;
+    [key: string]: any;
+}
+
+const ExportButton: FunctionComponent<Props> = ({
```

Comment:
As we pass the rest props to the Button, I would also add the MUI Button Props (even though we use our own button, it is based on the MUI one):

```ts
import { ButtonProps } from '@material-ui/core/Button';

const ExportButton: FunctionComponent<Props & ButtonProps> = ({
```